// @cryptostats/sdk/dist/libs/Ethers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Ethers = void 0;
const ethers_1 = require("ethers");
class Provider extends ethers_1.ethers.providers.JsonRpcProvider {
    constructor(url, id, archive, chainData) {
        super(url);
        this.isArchive = archive;
        this.id = id;
        this.chainData = chainData;
    }
    async _getBlockTag(blockTag) {
        blockTag = await blockTag;
        if (blockTag) {
            if (!this.isArchive) {
                throw new Error(`Unable to query by blockTag: provider is not archive node`);
            }
            if (/\d{4}-\d{2}-\d{2}/.test(blockTag.toString())) {
                blockTag = await this.chainData.getBlockNumber(blockTag, this.id);
            }
        }
        return super._getBlockTag(blockTag);
    }
}
const ERC20_ABI = [
    'function name() external view returns (string)',
    'function symbol() external view returns (string)',
    'function decimals() external view returns (uint8)',
    'function totalSupply() external view returns (uint256)',
    'function balanceOf(address account) external view returns (uint256)',
    'function transfer(address recipient, uint256 amount) external returns (bool)',
    'function allowance(address owner, address spender) external view returns (uint256)',
    'function approve(address spender, uint256 amount) external returns (bool)',
    'function transferFrom(address sender,address recipient,uint256 amount) external returns (bool)',
    'event Transfer(address indexed from, address indexed to, uint256 value)',
    'event Approval(address indexed owner, address indexed spender, uint256 value)',
];
class Ethers {
    constructor({ chainData }) {
        this.providersByNetwork = {};
        this.chainData = chainData;
        this.utils = ethers_1.utils;
        this.BigNumber = ethers_1.BigNumber;
        this.FixedNumber = ethers_1.FixedNumber;
    }
    addProvider(name, url, { archive = false, } = {}) {
        this.providersByNetwork[name] = { url, archive, provider: null };
    }
    getContract(address, abi, network = 'ethereum') {
        return new ethers_1.ethers.Contract(address, abi, this.getProvider(network));
    }
    getERC20Contract(address, network = 'ethereum') {
        return this.getContract(address, ERC20_ABI, network);
    }
    getProvider(network) {
        const providerData = this.providersByNetwork[network];
        if (!providerData) {
            throw new Error(`Network ${network} is not available`);
        }
        if (!providerData.provider) {
            providerData.provider = new Provider(providerData.url, network, providerData.archive, this.chainData);
        }
        return providerData.provider;
    }
}
exports.Ethers = Ethers;
